### PR TITLE
[cli] support -I {src-addr} for ping command

### PR DESCRIPTION
**Pull request by superwhd**

### CODE REVIEW

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -3252,49 +3252,61 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
 {
     otError            error = OT_ERROR_NONE;
     otPingSenderConfig config;
+    uint8_t argsLength = aArgsLength;
+    char **args = aArgs;
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3256]:
```suggestion
```
seems not necessary

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -3263,6 +3263,16 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
 
     memset(&config, 0, sizeof(config));
 
+    if (aArgsLength >= 2)
+    {
+        if (!strcmp(aArgs[0], "-I"))
+        {
+            SuccessOrExit(error = ParseAsIp6Address(aArgs[1], config.mSource));
```
Comment (bukepo) [inline src/cli/cli.cpp:3270]: Verify this address is an interface address?

**Review by bukepo — APPROVED** [review]

File: src/cli/cli.cpp [hunk 3]
```diff
@@ -3263,6 +3263,27 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
...
+            bool                  valid        = false;
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3271]: Should put all variables at the beginning of the scope.

See: https://github.com/openthread/openthread/blob/main/STYLE_GUIDE.md#white-space

File: src/cli/cli.cpp [hunk 4]
```diff
@@ -3263,6 +3263,27 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
...
+            const otNetifAddress *unicastAddrs = otIp6GetUnicastAddresses(mInstance);
+            for (const otNetifAddress *addr = unicastAddrs; addr; addr = addr->mNext)
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3273]:
```suggestion
            const otNetifAddress *unicastAddrs = otIp6GetUnicastAddresses(mInstance);

            for (const otNetifAddress *addr = unicastAddrs; addr; addr = addr->mNext)
```
nit: add space between variable declarations and statements

File: src/cli/cli.cpp [hunk 5]
```diff
@@ -3263,6 +3263,27 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
...
+            {
+                if (otIp6IsAddressEqual(&addr->mAddress, &config.mSource))
+                {
+                    valid = true;
+                    break;
+                }
+            }
+            VerifyOrExit(valid, error = OT_ERROR_INVALID_ARGS);
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3281]: Skip valid checking if REFERENCE_DEVICE == 1 ?
I think we may need to test packets with invalid source address.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

with nits

File: src/cli/cli.cpp [hunk 6]
```diff
@@ -3277,6 +3277,32 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
...
+                {
+                    valid = true;
+                    break;
+                }
+            }
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+            VerifyOrExit(valid, error = OT_ERROR_INVALID_ARGS);
+#else
+            OT_UNUSED_VARIABLE(valid);
+#endif
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3300]:
```suggestion
#if !OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
            bool                  valid        = false;
            const otNetifAddress *unicastAddrs = otIp6GetUnicastAddresses(mInstance);

            SuccessOrExit(error = ParseAsIp6Address(aArgs[1], config.mSource));
            for (const otNetifAddress *addr = unicastAddrs; addr; addr = addr->mNext)
            {
                if (otIp6IsAddressEqual(&addr->mAddress, &config.mSource))
                {
                    valid = true;
                    break;
                }
            }
            VerifyOrExit(valid, error = OT_ERROR_INVALID_ARGS);
#endif
```

**Review by jwhui — APPROVED** [review]

**Comment by jwhui** [comment]
@superwhd , a number of tests are still failing.

**Comment by superwhd** [comment]
Thanks @jwhui.  I'm taking a look at the failures.

File: script/test [hunk 7]
```diff
@@ -263,7 +263,18 @@ do_build_otbr_docker()
...
+        "-DOT_DNS_CLIENT=ON"
+        "-DOT_DUA=ON"
+        "-DOT_MLR=ON"
+        "-DOT_SLAAC=ON"
+        "-DOT_SRP_CLIENT=ON"
+        "-DOT_TREL=OFF"
+        "-DOTBR_DUA_ROUTING=ON"
+        "-DOTBR_REST=OFF"
+        "-DOTBR_WEB=OFF"
+    )
```
Comment (simonlingoogle) [inline script/test:277]:
```suggestion
```
Not include changes from another PR ? @superwhd

File: src/cli/cli.cpp [hunk 8]
```diff
@@ -3277,6 +3277,33 @@ otError Interpreter::ProcessPing(uint8_t aArgsLength, char *aArgs[])
...
+    {
+        if (!strcmp(aArgs[0], "-I"))
+        {
+            SuccessOrExit(error = ParseAsIp6Address(aArgs[1], config.mSource));
+#if !OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+            {
+                bool                  valid        = false;
+                const otNetifAddress *unicastAddrs = otIp6GetUnicastAddresses(mInstance);
+
+                SuccessOrExit(error = ParseAsIp6Address(aArgs[1], config.mSource));
```
Comment (simonlingoogle) [inline src/cli/cli.cpp:3290]:
```suggestion
```
why do it twice?

**Review by jwhui — APPROVED** [review]

**Pull request merged**